### ISSUE TITLE
Pro Controller connected with USB, doesn't reconnect after computer resumes from suspend

### ISSUE THREAD

**Issue by SuperSamus** [DanielOgorchock/linux#42]
Normally, if you connect a Nintendo Switch Pro Controller via USB, it is automatically connected when turning the computer on.
If however the computer is suspended, when resuming the computer, the controller won't reconnect, even if you press a button. The ways to reconnect it are either:

- Press the sync button on the controller;
- Unplug and plug the USB cable;
- Reboot the computer.

Relevant logs from `journalctl` when suspending:
```
lug 11 00:39:44 nixos kernel: nintendo 0003:057E:2009.000F: compensating for 37 dropped IMU reports
lug 11 00:39:44 nixos kernel: nintendo 0003:057E:2009.000F: delta=504 avg_delta=13
```

When resuming:
```
lug 11 00:39:44 nixos kernel: leds 0003:057E:2009.000F:green:player-1: Setting an LED's brightness failed (-110)
lug 11 00:39:45 nixos kernel: leds 0003:057E:2009.000F:green:player-2: Setting an LED's brightness failed (-110)
lug 11 00:39:45 nixos kernel: leds 0003:057E:2009.000F:green:player-3: Setting an LED's brightness failed (-110)
lug 11 00:39:46 nixos kernel: leds 0003:057E:2009.000F:green:player-4: Setting an LED's brightness failed (-110)
lug 11 00:39:46 nixos kernel: leds 0003:057E:2009.000F:blue:player-5: Setting an LED's brightness failed (-110)
```

After pressing the sync button on the controller (still plugged in via USB):
```
lug 11 00:49:21 nixos kernel: usb 3-2: USB disconnect, device number 17
lug 11 00:49:22 nixos kernel: usb 3-2: new full-speed USB device number 18 using xhci_hcd
lug 11 00:49:22 nixos kernel: usb 3-2: New USB device found, idVendor=057e, idProduct=2009, bcdDevice= 2.10
lug 11 00:49:22 nixos kernel: usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
lug 11 00:49:22 nixos kernel: usb 3-2: Product: Pro Controller
```
...etcetera.

**Comment by DanielOgorchock** [comment]
Thanks for pointing this out. Yeah the driver doesn't have pm hooks implemented. I should really add those.

**Comment by SuperSamus** [comment]
Hi, is there any update on this?

The template for should be mostly like this, right? (Shamelessly copied from [hid-sony](https://github.com/torvalds/linux/blob/master/drivers/hid/hid-sony.c#L2182-L2217))

<details>
<summary>Template</summary>

```c
#ifdef CONFIG_PM

static int nintendo_hid_suspend(struct hid_device *hdev, pm_message_t message)
{
#ifdef CONFIG_NINTENDO_FF

	/* On suspend stop any running force-feedback events (I don't know if it is necessary on Nintendo) */
	struct joycon_ctlr *ctlr = hid_get_drvdata(hdev);
        /* etc */

#endif
	return 0;
}

static int nintendo_hid_resume(struct hid_device *hdev)
{
	/*
	 * The controllers on USB need to be reinitialized
	 * on resume or they won't behave properly.
	 */
	if (hdev->bus == BUS_USB) {
		/* Instructions to resume */
	}

	return 0;
}

#endif

/* ... */

static struct hid_driver nintendo_hid_driver = {
	.name		= "nintendo",
	.id_table	= nintendo_hid_devices,
	.probe		= nintendo_hid_probe,
	.remove		= nintendo_hid_remove,
	.raw_event	= nintendo_hid_event,

#ifdef CONFIG_PM
	.suspend          = nintendo_hid_suspend,
	.resume	          = nintendo_hid_resume,
	.reset_resume     = nintendo_hid_resume,
#endif
};
```
</details>

**Comment by SuperSamus** [comment]
Well, I tried it myself.
After trying to compile the Linux kernel to load a module (first time), I finally did it.
I didn't test it much, but it seems to work well.
~Here is the patch: [pro-controller-suspend.txt](https://github.com/DanielOgorchock/linux/files/11996278/pro-controller-suspend.txt)~
EDIT: [pro-controller-suspend.txt](https://github.com/DanielOgorchock/linux/files/12022007/pro-controller-suspend.txt)


If no one has anything against it, I'll send it after a week. (My first kernel contribution)
